docs/spark.rst

```diff
@@ -0,0 +1,106 @@
+the cluster configuration.
+
+We decompose this cluster into two separate classes inside of Toil:
+:class:`toil.lib.spark.SparkService` and
+:class:`toil.lib.spark.WorkerService`. The
+:class:`toil.lib.spark.SparkService` class launches the Spark master
+and HDFS namenode, and then adds one or more
+:class:`toil.lib.spark.WorkerService` child services. The
+:class:`toil.lib.spark.WorkerService` class then launches the
+Spark worker and HDFS datanode. The individual Spark and HDFS applications are
+run using docker containers from cgl-docker-lib_.
+
+.. _cgl-docker-lib: https://github.com/BD2KGenomics/cgl-docker-lib
+
+Launching a Spark Cluster
+-------------------------
+
+To launch a Spark cluster, import:
+
+::
+        from toil_lib.spark import spawn_spark_cluster
+
+Then, add a service job:
+
+::
+        masterIPPromise = spawn_spark_cluster(job, ...)
+
+The service job will return a promise for the IP address of the master node.
+The service job takes several parameters:
+
+- job: The job that the Spark cluster should descend from.
+- sudo: Whether or not to run the Docker containers with sudo permission.
+- numWorkers: The number of workers to start in the cluster. Must be greater than
+  or equal to one.
+
+We also take several optional parameters:
+
+- cores: An optional parameter to set the number of cores that each machine should
+  use. If not provided, we will auto-detect the cores on a worker.
+- overrideLeaderIP: an optional parameter that can override the leader IP.
+- memory: The amount of memory to allocate for each worker used by Spark.
+- disk: The amount of disk to request for each worker.
+
+When launching a Spark cluster, there is one important caveat: the job that
+launches the services must be executed with `checkpoint = True`. If the job is
+not checkpointable, then the cluster may not restart robustly. This is important
+if you are running on a preemptable cluster: in Spark and HDFS, failed workers can
+be restarted, but a failed master will cause all workers in the cluster to
+dissociate. If the cluster does not restart, this causes a deadlock.
+
+Using Locally
+-------------
+
+To use Spark locally, you must install Toil, toil-lib, and Docker. Docker installation
+instructions can be found [here](https://docs.docker.com/engine/installation/).
+For examples of how to connect to a Toil-based Spark cluster, see the Spark
+unit tests:
+
+::
+        make test tests=src/toil_lib/test/test_spark.py
+
+These unit tests create a Spark cluster, and then run a simple job using PySpark.
+
+You must install Spark and make PySpark available on your
+Python PATH. Pre-built binaries for Spark can be downloaded from the
+[Apache Spark downloads page](http://spark.apache.org/downloads.html). Once
+these are downloaded, untar the archive, and add the following location
+to your PYTHONPATH environment variable:
+
+::
+        export PYTHONPATH=${SPARK_HOME}/python:${SPARK_HOME}/python/lib/py4j-0.10.1-src.zip:${PYTHONPATH}
+
```

Comment:
`spark.rst` is being moved over from the Toil docs. @fnothaft is there a reason for this specific link?